### ISSUE TITLE
Checkout reference

### ISSUE THREAD

**Issue by catayoyo** [underlandxyz/fb-sprint-1#1]
Hi there,

Rosemary told me that for task 4 I needed to create a new Git branch called YOURUSERNAME-sprint-1 and checkout that branch so I'm working on the branch. I thought I had done this but it doesn't seem to be the case and I can't work out why?

I emailed the tech team via the help email as I got errors when I tried to push back to the repository as well. I got errors about AOuth tokens. The same happened when I checked on Sourcetree so I'm not sure if it's a problem with having a free account or something either.

**Comment by stdlibdoh** [comment]
Hiya,

I've just emailed you about the errors you're facing when pushing, so hopefully that should be fine now.

So, if I understand it correctly you've:

* Created a branch by running > git branch <branch_name>

* Checked out that branch by running > git checkout <branch_name>

If that's done, if you run > git branch, you should see all the branches you've created, with a star symbol next to the one you're currently on.

Can you please confirm that?

**Comment by catayoyo** [comment]
Hi,

I'm only using the drop down menus which gives me "checkout reference" and when i try and type the branch in it says that can't be used. 

Clicking on branch as you said gives me this: https://i.imgur.com/o0Y3EnG.png but I'm completely new to git and not actually sure if that's what you mean at all, sorry.

**Comment by stdlibdoh** [comment]
Oh I see. I've just entered your workspace and think I get it now. No need to apologise, we all go through this at one stage or another (I still remember mine vividly, cue cold sweats whilst screaming at git to work)

I'm just going to post something a bit detailed (and apologies in advance if it sounds too extra)

A branch effectively creates a snapshot of your code so any work that you do while checked out (within a) branch, is purely on that branch. If you checkout (go into) another branch, whatever work you created is gone. I'm going to assume that all your work would've been created in your catayoyo-sprint-1 branch. If it's not, let me know and I'll adjust these instructions.

I can see you've forked and cloned the repository to your personal GitHub profile. Next step is adding all your changes to it.

* You should see 3 tabs at the bottom of the screen: 'dev-machine', 'Terminal', and 'db'. We just need Terminal so click on that

* If you look at the command prompt, you'll see where you are right now which should be the roof of the directory aka user@<long_number>:/projects . We need to go into our directory so, again on the Terminal, write:
```cd catayoyo-sprint-1``` (as that's the name of the folder we're trying to get into)

* if you run ```git status``` you can see all the files that were edited. If that output makes sense, add them by running ```git add . ```. That final dot is important as it will add everything in every folder

* Next we're going to commit our changes so run ```git commit -m "<any message you think is appropriate>"```

* Finally we're going to push our changes to your repository so run ```git push origin <your_branch_name>```

* Note: The only branch I see is one called "catayoyo", you might've been trying to push from a branch that doesn't exist

**Comment by catayoyo** [comment]
Hello,

Not extra at all, I definitely needed it breaking down more than before as I thought we had to use the menus/i don't know how to use the commands properly yet. Sourcetree, which my company uses, is all clicking rather than using cmd prompts. 

I have an issue in that my email/username is in my preferences, yet when I do the commit step it says my name isn't in there and asks me to config, yet when i follow the steps i get this message:

`*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.

fatal: empty ident name (for <>) not allowed`

**Comment by stdlibdoh** [comment]
Ah beautiful! We're getting somewhere now! 🎉 

That's expected. Pretty sure it has something to do with Che not fetching credentials automatically but it's something we can fix.

Follow those instructions as detailed:

``` git config --global user.email "amelia.robertson@aviva.com"```
``` git config --global user.name "Amelia Robertson"```

After that, just revert back to what you were doing. If I remember correctly, that is thrown when you try to commit. Commit again, push as detailed in the previous post and 🤞 it should be fine.

After that, there's the Pull Request but we'll cross that bridge when we have to.

**Comment by catayoyo** [comment]
Hmm... I did that last time and with catayoyo as the user name but it gave me the empty indent name again just now as well.

Any ideas why?

**Comment by stdlibdoh** [comment]
Bear with me for a moment, I've never encountered that before, will need to look into it

**Comment by catayoyo** [comment]
Sure, thanks for the help! https://i.imgur.com/g1ppOPW.png There's a screenshot of the terminal in case it gives extra insight.

**Comment by stdlibdoh** [comment]
Wow that was interesting! Let me try to give you a full rundown of what I did.

So that error message was indeed appearing because the user.email and user.name weren't set. But we set them so what gives??

Under the hood, any git repository has a few configuration files that tell it exactly how to behave. The folder is called .git (the dot before git indicates that it's a hidden file). Going into it reveals a few different files but the one we're looking after is called config. I was in your workspace trying to figure it out so I'll post the contents of it. Also, to get it yourself, just run:

```
git config --list
```

```
core.excludesfile=/home/user/.gitignore_codenvy
user.name=Amelia Robertson
user.email=amelia.robertson@aviva.com
core.repositoryformatversion=0
core.filemode=true
core.logallrefupdates=true
remote.origin.url=https://github.com/catayoyo/fb-sprint-1
remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
branch.master.remote=origin
branch.master.merge=refs/heads/master
user.name=
user.email=
```

We have two sets of user.email, user.name. A quick google search took me to this [StackOverflow](https://stackoverflow.com/questions/4310974/more-than-one-value-for-the-key-user-name-git) post. TLDR is that it's always looking at the closest config entries to the end of the file. It was going through the list and, since the bottom ones aren't set, it was simply not accepting anything.

In the post I mentioned, one entry simply states to overwrite all entries and that it should work. So I did it on your terminal. I ran:

```git config --replace-all user.email``` or ```user.name``` and set the values accordingly.

I did a few adding and committing to test it out so in the end, I had to reset everything. This just means that when you're ready, go through the add, commit, push flow from the beginning.

[More on reset](https://git-scm.com/book/en/v2/Git-Tools-Reset-Demystified)

One more thing I noticed in your screenshots:

It is entirely my fault but the <> was simply to indicate this was a value to be changed. So when you're running any kind of commands, there isn't any need to include them (and it's best if you dont). So the command I can see:

```
git add <answers/catayoyo/FizzBuzz/FizzBuzz.java>
```
Would simply be:
```
git add answers/catayoyo/FizzBuzz/FizzBuzz.java
```

Same for the commit message. The m flag (which is what we call when we use -m) is simply just saying that, in our commit command, we're also adding a message. So simply running

```
git commit -m "I've now committed something"
```
is more than enough.

TIL something new about Git ! 🌈
